packages/firestore/src/core/bundle.ts

```diff
@@ -98,3 +107,132 @@ export class BundleConverter {
+    totalBytes: metadata.totalBytes!
+  };
+}
+
+export class LoadResult {
+  constructor(
+    readonly progress: firestore.LoadBundleTaskProgress,
+    readonly changedDocs?: MaybeDocumentMap
+  ) {}
+}
+
+/**
+ * A class to process the elements from a bundle, load them into local
+ * storage and provide progress update while loading.
+ */
+export class BundleLoader {
+  /** The current progress of loading */
+  private progress: firestore.LoadBundleTaskProgress;
+  /** Batched queries to be saved into storage */
+  private queries: bundleProto.NamedQuery[] = [];
+  /** Batched documents to be saved into storage */
+  private documents: BundledDocuments = [];
+
+  constructor(
+    private metadata: bundleProto.BundleMetadata,
+    private localStore: LocalStore
+  ) {
+    this.progress = bundleInitialProgress(metadata);
+  }
+
+  /**
+   * Adds an element from the bundle to the loader.
+   *
+   * Returns a new progress if adding the element leads to a new progress,
+   * otherwise returns null.
+   */
+  addSizedElement(
+    element: SizedBundleElement
+  ): firestore.LoadBundleTaskProgress | null {
+    debugAssert(!element.isBundleMetadata(), 'Unexpected bundle metadata.');
+
+    this.progress.bytesLoaded += element.byteLength;
+
+    let documentsLoaded = this.progress.documentsLoaded;
+
+    if (element.payload.namedQuery) {
+      this.queries.push(element.payload.namedQuery);
+    } else if (element.payload.documentMetadata) {
+      this.documents.push({ metadata: element.payload.documentMetadata });
+      if (!element.payload.documentMetadata.exists) {
+        ++documentsLoaded;
+      }
+    } else if (element.payload.document) {
+      debugAssert(
+        this.documents.length > 0 &&
+          this.documents[this.documents.length - 1].metadata.name ===
+            element.payload.document.name,
+        'The document being added does not match the stored metadata.'
+      );
+      this.documents[this.documents.length - 1].document =
+        element.payload.document;
+      ++documentsLoaded;
+    }
+
+    if (documentsLoaded !== this.progress.documentsLoaded) {
+      this.progress.documentsLoaded = documentsLoaded;
+      return { ...this.progress };
+    }
+
+    return null;
+  }
+
+  /**
+   * Update the progress to 'Success' and return the updated progress.
+   */
+  async complete(): Promise<LoadResult> {
+    const lastDocument =
+      this.documents.length === 0
+        ? null
+        : this.documents[this.documents.length - 1];
+    debugAssert(
+      !!lastDocument ||
+        !lastDocument!.metadata.exists ||
+        (!!lastDocument!.metadata.exists && !!lastDocument!.document),
+      'Bundled documents ends with a document metadata.'
+    );
+
+    for (const q of this.queries) {
+      await saveNamedQuery(this.localStore, q);
+    }
+
+    let changedDocs;
+    if (this.documents.length > 0) {
+      changedDocs = await applyBundleDocuments(this.localStore, this.documents);
```

Comment:
How often do you think bundles will be empty? If this is rare, then we could just call `applyBundleDocuments`, which only has a little bit of overhead but removes two if statements from your code. You already don't call into this code for skipped bundles, so I would think `changedDocs` would rarely be empty.